File: sources/net.sf.j2s.core/src/net/sf/j2s/core/utils/CorePluginConsole.java
package net.sf.j2s.core.utils;

import java.io.Console;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Calendar;
import java.util.logging.Formatter;
import java.util.logging.Level;
import java.util.logging.LogRecord;

import org.eclipse.swt.graphics.Color;
import org.eclipse.swt.widgets.Display;
import org.eclipse.ui.console.ConsolePlugin;
import org.eclipse.ui.console.IConsole;
import org.eclipse.ui.console.IConsoleManager;
import org.eclipse.ui.console.MessageConsole;
import org.eclipse.ui.console.MessageConsoleStream;

public class CorePluginConsole {
	/**
	 * Message console for displaying J2sCorePlugin runtime message
	 */
	private static MessageConsole sConsole = null;
	/**
	 * Successful console stream (text in green)
	 */
	private static MessageConsoleStream sSuccessStream;
	/**
	 * Normal console stream (text in black)
	 */
	private static MessageConsoleStream sNormalStream;
	/**
	 * Error console stream (text in red)
	 */
	private static MessageConsoleStream sErrorStream;
	/**
	 * Color red
	 */
	private static Color sColorRed;
    /**
     * Color green
     */
    private static Color sColorGreen;
	/**
	 * Message formatter
	 */
	private static Formatter sFormatter;
	

	/**
	 * get MptPluginConsole from default ConsoleManager to init CorePluginConsole
	 */
	public static void initConsoleView() {
		IConsoleManager consoleManager = ConsolePlugin.getDefault().getConsoleManager();
		IConsole[] IConsole = consoleManager.getConsoles();
		for(int i = 0; i != IConsole.length;i++){
			if(IConsole[i].getName().equals(MptConstants.Mayloon_Console_Name)){
				sConsole = (MessageConsole) IConsole[i];
			}
		}
		if(sConsole == null){
			sConsole = new MessageConsole(MptConstants.Mayloon_Console_Name, null);
		}
		sSuccessStream = sConsole.newMessageStream();
		sNormalStream = sConsole.newMessageStream();
		sErrorStream = sConsole.newMessageStream();
		Display display = ConsolePlugin.getDefault().getWorkbench()
				.getDisplay();
		sColorRed = new Color(display, 0xFF, 0x00, 0x00);
		sColorGreen = new Color(display, 0x00, 0xFF, 0x00);
		display.asyncExec(new Runnable() {
			public void run() {
				sErrorStream.setColor(sColorRed);
				sSuccessStream.setColor(sColorGreen);
			}
		});
		sFormatter = new Formatter() {
			public String format(LogRecord record) {
				StringBuffer buffer = new StringBuffer();
				buffer.append(String.format(
						"[%1$tF_%1$tT:%1$tL] [%2$s] [%3$s] %4$s\n",
						Calendar.getInstance(), record.getLevel(),
						record.getLoggerName(), record.getMessage()));
				if (record.getThrown() != null) {
					StringWriter writer = new StringWriter();
					record.getThrown().printStackTrace(new PrintWriter(writer));
					buffer.append(writer.toString());
				}
				return buffer.toString();
			}
		};
	}


/**
 * Print general message to normal console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public static void general(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.INFO, String.format(format, args));
	record.setLoggerName(tag);
	sNormalStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}

/**
 * Print successful message to normal console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public static void success(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.INFO, String.format(format, args));
	record.setLoggerName(tag);
	sSuccessStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}

/**
 * Print warning message to error console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public static void warning(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.WARNING, String.format(format, args));
	record.setLoggerName(tag);
	sErrorStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}

/**
 * Print error message to error console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public static void error(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.SEVERE, String.format(format, args));
	record.setLoggerName(tag);
	sErrorStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}
/**
 * Return normal console stream
 * @return MessageConsoleStream
 */
public static MessageConsoleStream getNormalStream(){
	return sNormalStream;
}

/**
 * Return error console stream
 * @return MessageConsoleStream
 */
public static MessageConsoleStream getErrorStream(){
	return sErrorStream;
}

/**
 * @return the sSuccessStream
 */
public static MessageConsoleStream getsSuccessStream() {
	return sSuccessStream;
}

/**
 * Return message formatter
 * @return Formatter
 */
public static Formatter getFormatter(){
	return sFormatter;
}

/**
 * Destroy
 */
public static void destroy(){
	sColorRed.dispose();
	sColorGreen.dispose();
	try {
		sNormalStream.close();
	} catch (IOException e) {
	}
	try {
		sErrorStream.close();
	} catch (IOException e) {
	}
	try {
		sSuccessStream.close();
	} catch (IOException e) {
	}
}

public static boolean isInit(){
	return sConsole==null ? false : true;
} 
}

File: sources/net.sf.j2s.core/bin/net/sf/j2s/core/utils/CorePluginConsole.js
Clazz.declarePackage ("net.sf.j2s.core.utils");
Clazz.load (null, "net.sf.j2s.core.utils.CorePluginConsole", ["java.io.PrintWriter", "$.StringWriter", "java.lang.StringBuffer", "java.util.Calendar", "java.util.logging.Formatter", "$.Level", "$.LogRecord", "net.sf.j2s.core.utils.MptMessages", "$wt.graphics.Color", "org.eclipse.ui.console.ConsolePlugin"], function () {
c$ = Clazz.declareType (net.sf.j2s.core.utils, "CorePluginConsole");
c$.initConsoleView = Clazz.defineMethod (c$, "initConsoleView", 
function () {
var consoleManager = org.eclipse.ui.console.ConsolePlugin.getDefault ().getConsoleManager ();
var IConsole = consoleManager.getConsoles ();
for (var i = 0; i != IConsole.length; i++) {
if (IConsole[i].getName ().equals (net.sf.j2s.core.utils.MptMessages.Mayloon_Console_Name)) {
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sConsole = IConsole[i], net.sf.j2s.core.utils.CorePluginConsole.prototype.sConsole = net.sf.j2s.core.utils.CorePluginConsole.sConsole, $t$);
}}
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream = net.sf.j2s.core.utils.CorePluginConsole.sConsole.newMessageStream (), net.sf.j2s.core.utils.CorePluginConsole.prototype.sSuccessStream = net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream, $t$);
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sNormalStream = net.sf.j2s.core.utils.CorePluginConsole.sConsole.newMessageStream (), net.sf.j2s.core.utils.CorePluginConsole.prototype.sNormalStream = net.sf.j2s.core.utils.CorePluginConsole.sNormalStream, $t$);
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sErrorStream = net.sf.j2s.core.utils.CorePluginConsole.sConsole.newMessageStream (), net.sf.j2s.core.utils.CorePluginConsole.prototype.sErrorStream = net.sf.j2s.core.utils.CorePluginConsole.sErrorStream, $t$);
var display = org.eclipse.ui.console.ConsolePlugin.getDefault ().getWorkbench ().getDisplay ();
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sColorRed =  new $wt.graphics.Color (display, 0xFF, 0x00, 0x00), net.sf.j2s.core.utils.CorePluginConsole.prototype.sColorRed = net.sf.j2s.core.utils.CorePluginConsole.sColorRed, $t$);
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sColorGreen =  new $wt.graphics.Color (display, 0x00, 0xFF, 0x00), net.sf.j2s.core.utils.CorePluginConsole.prototype.sColorGreen = net.sf.j2s.core.utils.CorePluginConsole.sColorGreen, $t$);
display.asyncExec (((Clazz.isClassDefined ("net.sf.j2s.core.utils.CorePluginConsole$1") ? 0 : net.sf.j2s.core.utils.CorePluginConsole.$CorePluginConsole$1$ ()), Clazz.innerTypeInstance (net.sf.j2s.core.utils.CorePluginConsole$1, this, null)));
($t$ = net.sf.j2s.core.utils.CorePluginConsole.sFormatter = ((Clazz.isClassDefined ("net.sf.j2s.core.utils.CorePluginConsole$2") ? 0 : net.sf.j2s.core.utils.CorePluginConsole.$CorePluginConsole$2$ ()), Clazz.innerTypeInstance (net.sf.j2s.core.utils.CorePluginConsole$2, this, null)), net.sf.j2s.core.utils.CorePluginConsole.prototype.sFormatter = net.sf.j2s.core.utils.CorePluginConsole.sFormatter, $t$);
});
c$.general = Clazz.defineMethod (c$, "general", 
function (tag, format, args) {
var record =  new java.util.logging.LogRecord (java.util.logging.Level.INFO, String.format (format, [args]));
record.setLoggerName (tag);
net.sf.j2s.core.utils.CorePluginConsole.sNormalStream.print (net.sf.j2s.core.utils.CorePluginConsole.sFormatter.format (record));
org.eclipse.ui.console.ConsolePlugin.getDefault ().getConsoleManager ().showConsoleView (net.sf.j2s.core.utils.CorePluginConsole.sConsole);
}, "~S,~S,~A");
c$.success = Clazz.defineMethod (c$, "success", 
function (tag, format, args) {
var record =  new java.util.logging.LogRecord (java.util.logging.Level.INFO, String.format (format, [args]));
record.setLoggerName (tag);
net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream.print (net.sf.j2s.core.utils.CorePluginConsole.sFormatter.format (record));
org.eclipse.ui.console.ConsolePlugin.getDefault ().getConsoleManager ().showConsoleView (net.sf.j2s.core.utils.CorePluginConsole.sConsole);
}, "~S,~S,~A");
c$.warning = Clazz.defineMethod (c$, "warning", 
function (tag, format, args) {
var record =  new java.util.logging.LogRecord (java.util.logging.Level.WARNING, String.format (format, [args]));
record.setLoggerName (tag);
net.sf.j2s.core.utils.CorePluginConsole.sErrorStream.print (net.sf.j2s.core.utils.CorePluginConsole.sFormatter.format (record));
org.eclipse.ui.console.ConsolePlugin.getDefault ().getConsoleManager ().showConsoleView (net.sf.j2s.core.utils.CorePluginConsole.sConsole);
}, "~S,~S,~A");
c$.error = Clazz.defineMethod (c$, "error", 
function (tag, format, args) {
var record =  new java.util.logging.LogRecord (java.util.logging.Level.SEVERE, String.format (format, [args]));
record.setLoggerName (tag);
net.sf.j2s.core.utils.CorePluginConsole.sErrorStream.print (net.sf.j2s.core.utils.CorePluginConsole.sFormatter.format (record));
org.eclipse.ui.console.ConsolePlugin.getDefault ().getConsoleManager ().showConsoleView (net.sf.j2s.core.utils.CorePluginConsole.sConsole);
}, "~S,~S,~A");
c$.getNormalStream = Clazz.defineMethod (c$, "getNormalStream", 
function () {
return net.sf.j2s.core.utils.CorePluginConsole.sNormalStream;
});
c$.getErrorStream = Clazz.defineMethod (c$, "getErrorStream", 
function () {
return net.sf.j2s.core.utils.CorePluginConsole.sErrorStream;
});
c$.getsSuccessStream = Clazz.defineMethod (c$, "getsSuccessStream", 
function () {
return net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream;
});
c$.getFormatter = Clazz.defineMethod (c$, "getFormatter", 
function () {
return net.sf.j2s.core.utils.CorePluginConsole.sFormatter;
});
c$.destroy = Clazz.defineMethod (c$, "destroy", 
function () {
net.sf.j2s.core.utils.CorePluginConsole.sColorRed.dispose ();
net.sf.j2s.core.utils.CorePluginConsole.sColorGreen.dispose ();
try {
net.sf.j2s.core.utils.CorePluginConsole.sNormalStream.close ();
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
try {
net.sf.j2s.core.utils.CorePluginConsole.sErrorStream.close ();
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
try {
net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream.close ();
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
});
c$.$CorePluginConsole$1$ = function () {
Clazz.pu$h ();
c$ = Clazz.declareAnonymous (net.sf.j2s.core.utils, "CorePluginConsole$1", null, Runnable);
Clazz.overrideMethod (c$, "run", 
function () {
net.sf.j2s.core.utils.CorePluginConsole.sErrorStream.setColor (net.sf.j2s.core.utils.CorePluginConsole.sColorRed);
net.sf.j2s.core.utils.CorePluginConsole.sSuccessStream.setColor (net.sf.j2s.core.utils.CorePluginConsole.sColorGreen);
});
c$ = Clazz.p0p ();
};
c$.$CorePluginConsole$2$ = function () {
Clazz.pu$h ();
c$ = Clazz.declareAnonymous (net.sf.j2s.core.utils, "CorePluginConsole$2", java.util.logging.Formatter);
Clazz.defineMethod (c$, "format", 
function (record) {
var buffer =  new StringBuffer ();
buffer.append (String.format ("[%1$tF_%1$tT:%1$tL] [%2$s] [%3$s] %4$s\n", [java.util.Calendar.getInstance (), record.getLevel (), record.getLoggerName (), record.getMessage ()]));
if (record.getThrown () != null) {
var writer =  new java.io.StringWriter ();
record.getThrown ().printStackTrace ( new java.io.PrintWriter (writer));
buffer.append (writer.toString ());
}return buffer.toString ();
}, "java.util.logging.LogRecord");
c$ = Clazz.p0p ();
};
Clazz.defineStatics (c$,
"sConsole", null,
"sSuccessStream", null,
"sNormalStream", null,
"sErrorStream", null,
"sColorRed", null,
"sColorGreen", null,
"sFormatter", null);
});
